### ISSUE TITLE
Request for Timeline on Release of Training and Inference Scripts

### ISSUE THREAD

**Issue by xiaogengen-007** [BailiangJ/rethink-reg#1]
Hi BailiangJ,

Thank you for your excellent work and for sharing the code! I’m really excited to test your training strategy on our dataset.

Could you please share an approximate timeline for when the training and inference scripts might be released?

Thank you!

**Comment by xiaogengen-007** [comment]
Hi @BailiangJ,

Thank you for sharing the code—it’s been incredibly helpful!

I’m also particularly interested in the high-resolution deformation grid plot in your paper. I am wondering how you were able to show the deformed grid with such high resolution. Could you share the code you used to generate it, or provide some guidance on how to reproduce that effect?

Thank you so much in advance!

**Comment by BailiangJ** [comment]
Hi @xiaogengen-007 , 

Yes sure, here is the code for the deformation evaluation.

It is basically generated with the help of ChatGPT and Claude.

The white background and red line might not be optimal for a good contrast. Some tuning might be required.

You are also welcome to share your optimized visualization code so others can benefit from it.

Cheers,
Bailiang
```
import os
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection
from scipy.ndimage import map_coordinates
import matplotlib.colors as mcolors

def plot_grid(x, y,
              save_name: str,
              figsize=(10, 8),
              rgb=None,
              **kwargs):
    figsize = (figsize[1], figsize[0])
    fig, ax = plt.subplots(figsize=figsize)

    # print(x.shape, x.T.shape)
    x = x.T
    y = y.T
    segs1 = np.stack((y, -x), axis=2)

    # segs1 = np.stack((x, y), axis=2)
    segs2 = segs1.transpose(1, 0, 2)

    ax.add_collection(LineCollection(segs1, **kwargs))
    ax.add_collection(LineCollection(segs2, **kwargs))
    ax.axis('off')
    ax.autoscale()
    # fig.savefig(save_name, dpi=300, bbox_inches='tight')
    return fig,ax

def plot_deformed_grid(disp: np.ndarray,
                       select_dim: int,
                       select_slice: int,
                       # img_shape: tuple,
                       step_size: int,
                       save_name: str,
                       figsize=(10,8)):
    '''
    Args:
        disp: displacement field in image space (**not** normalized to [-1, 1]), shape (H, W, D, 3)
        select_dim: selected dimension.
        select_slice: selected slice.
        step_size:
        save_name:
        figsize:

    Returns:

    '''
    dims = [0, 1, 2]
    dims.pop(select_dim)
    # height, width = disp.shape[dims[0]], disp.shape[dims[1]]
    # aspect_ratio = width / height
    # print(aspect_ratio, fig_height*width/height)
    # fig_width = fig_height * aspect_ratio

    select_slice = select_slice // step_size
    # set indexing='ij' whenever using np.meshgrid
    img_shape = disp.shape[:3]
    grid_x, grid_y, grid_z = np.meshgrid(
        np.arange(0, img_shape[0] + 1, step_size),
        np.arange(0, img_shape[1] + 1, step_size),
        np.arange(0, img_shape[2] + 1, step_size),
        indexing='ij')

    id_grids = [grid_x, grid_y, grid_z]

    disp_x = map_coordinates(disp[..., 0], (grid_x, grid_y, grid_z), order=0)
    disp_y = map_coordinates(disp[..., 1], (grid_x, grid_y, grid_z), order=0)
    disp_z = map_coordinates(disp[..., 2], (grid_x, grid_y, grid_z), order=0)
    disp_fields = [disp_x, disp_y, disp_z]

    def_grid = []
    for i in dims:
        def_grid.append(np.take(id_grids[i],
                                indices=select_slice,
                                axis=select_dim) + np.take(disp_fields[i],
                                                           indices=select_slice,
                                                           axis=select_dim))

    fig, ax = plot_grid(def_grid[0], def_grid[1], rgb=None, save_name=save_name,
                        figsize=figsize,
                        color="red")
    fig.savefig(save_name, dpi=300, bbox_inches='tight')
```

**Comment by xiaogengen-007** [comment]
Hi @@BailiangJ ,

Thank you for the code! I played with it and got some nice visualizations.

I would like to share with you some of my visualizations and would also appreciate if you have any thoughts on it.

So I generated the following grid plot based on your code (using OASIS and the conventional greedy registration method). Below shows the same field in axial view with different colors and step size.

![Image](https://github.com/user-attachments/assets/33ca6374-99cf-40c9-a5bb-926e9e349c60)

![Image](https://github.com/user-attachments/assets/0cf8165e-9004-493d-9353-ff001012ba7c)

I noticed you used the forward deformation (sorry for the inaccurate naming here), which shows how the coordinates are mapped from the fixed image to the moving image. I believe ITK-SNAP also uses this convention (see image below):

![Image](https://github.com/user-attachments/assets/dcfe19b1-333d-485f-9a81-0c9337c165ce)

I am personally more familiar with the backward deformation grid, where a grid image is deformed similarly as warping the moving image. I think both VoxelMorph and TransMorph used such visualization. However, the grid width they used was relatively large (~8). I managed to get fine grids by upsampling the deformation field (this is indeed computation and memory consuming) and got the following backward grids:

![Image](https://github.com/user-attachments/assets/9adbda89-dc15-4b02-80d6-98a46173d913)


I am not sure if there is a consensus in the community on which way of grid visualization is preferred. Is there any particular reason that you chose the forward grid?

My initial thought is that they can be helpful in different situations.
On one hand, it would make more sense to overlay the backward grid with deformed images as they undergo the same warping/resampling process, and it would help spot which part of the moving image undergo implausible deformations:

![Image](https://github.com/user-attachments/assets/3de38ab3-ff9d-423a-a77a-c0b4fd230578)

On the other hand, the forward grid is more straightforward to calculate and goes naturally with the Jacobian determinant (and thus it can help us spot folding):

![Image](https://github.com/user-attachments/assets/dd0ad798-a39c-4c8c-9e1d-ee6477a80474)

**Comment by BailiangJ** [comment]
Hi @xiaogengen-007 ,

From my knowledge of image warping, most existing methods (torch.grid_sample, ANTs, greedy, Elastix, NiftyReg) use the image interpolation in Euler frames, formulated as:

$\text{Moved}(i,j) = \text{Moving}(\phi^{-1}(i,j))$

The deformation field $\phi^{-1}$ is mapping from fixed/target image space to moving/source image space, which is the actual output of all these methods. ITK-SNAP is using this convention because we are loading the deformation $\phi^{-1}$.

Therefore you are correct that it is actually deforming the fixed image grid instead of the moving image grid.

But the problem is how do you get the deformation $\phi$, by flipping the input order of the fixed and moving images? But this doesn't guarantee that it is the actual inverse of $\phi^{-1}$. And the moved image is actually generated by $\phi^{-1}$, not $\phi$.

This is quite tricky. From my personal point of view, the deformed grid visualization is just for a method-wise comparison and is usually less informative (especially for cross-sectional brain registration where too many details are deforming).  

I think plotting the Jdet actually makes more sense. The forward grid overlaid on Jdet (step size 4 and 2) look good.

**Comment by BailiangJ** [comment]
Hi @xiaogengen-007 ,

Correct me if I am wrong.

The two visualization methods are

1. "forward": deformed the _grid lines_ with the displacement (no image warping happened here)
2. "backward": create a **grid (checkerboard) image**, and then warp the **grid image** with the deformation field -> the **grid image** undergoes the same image warping process as the deformed moving image did

The _grid lines_ and **grid image** are two different things here.

Then, it makes sense that they can be "forward" and "backward" visualizations and the correspondence of shrinkage and expansion.

One of the downsides of the "backward" visualization is that the interpolation of the warping might distort the **grid image**; the lines in the **grid image** are also shrunk or expanded. But this is trivial.

Thanks a lot for the discussion! :)

**Comment by xiaogengen-007** [comment]
Yes! That's exactly what I mean, and I love your nice differentiation of _grid lines_ and grid image.

Thank you again for your patient answers and great comments!
